bridges/snowbridge/primitives/core/src/location.rs

```diff
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+//! # Location
+//!
+//! Location helpers for dealing with Tokens and Agents
+
+pub use polkadot_parachain_primitives::primitives::{
+	Id as ParaId, IsSystem, Sibling as SiblingParaId,
+};
+pub use sp_core::U256;
+
+use codec::Encode;
+use sp_core::H256;
+use sp_std::prelude::*;
+use xcm::prelude::{
+	AccountId32, AccountKey20, GeneralIndex, GeneralKey, GlobalConsensus, Location, PalletInstance,
+};
+use xcm_builder::{DescribeAllTerminal, DescribeFamily, DescribeLocation, HashedDescription};
+
+pub type AgentId = H256;
+
+/// Creates an AgentId from a Location. An AgentId is a unique mapping to a Agent contract on
+/// Ethereum which acts as the sovereign account for the Location.
+pub type AgentIdOf =
+	HashedDescription<AgentId, (DescribeHere, DescribeFamily<DescribeAllTerminal>)>;
+
+pub type TokenId = H256;
+
+/// Convert a token location to a stable ID that can be used on the Ethereum side
+pub type TokenIdOf = HashedDescription<
+	TokenId,
+	DescribeGlobalPrefix<(DescribeHere, DescribeFamily<DescribeTokenTerminal>)>,
+>;
+
+pub struct DescribeHere;
+impl DescribeLocation for DescribeHere {
+	fn describe_location(l: &Location) -> Option<Vec<u8>> {
+		match l.unpack() {
+			(0, []) => Some(Vec::<u8>::new().encode()),
+			_ => None,
+		}
+	}
+}
+pub struct DescribeGlobalPrefix<DescribeInterior>(sp_std::marker::PhantomData<DescribeInterior>);
+impl<Suffix: DescribeLocation> DescribeLocation for DescribeGlobalPrefix<Suffix> {
+	fn describe_location(l: &Location) -> Option<Vec<u8>> {
+		match (l.parent_count(), l.first_interior()) {
+			(_, Some(GlobalConsensus(network))) => {
```

Comment:
Are we allowing asset from a different consensus here?

Reply:
So I removed this because i am not sure if it matters for the purpose of token id generation the level of parents. But maybe you are correct in that it will always be 1.

Reply:
https://github.com/paritytech/polkadot-sdk/pull/5546/commits/34c655ba7bf8192b417f76e19be3d7d2435a3fcf